### PR TITLE
Create core Agent unit tests

### PR DESCRIPTION
**Pull request by BrianJKoopman**
## Description

This creates unit tests for the Tasks and Processes for the core OCS Agents, with the exception of the Host Master Agent. We also make some changes to the coverage configuration, workflows, and test markers, similar to what was recently done on https://github.com/simonsobs/socs/pull/196. Those changes include:
- Adding a 'run_once' parameter to existing Processes. This allows us to exit the Process in testing, just executing the loop once.
- Splitting out parts of the `.coveragerc` file, as some parts are relevant for combining coverage in the CI pipeline and others are needed for the tests when running within `./tests/`.
- Creating the spt3g marker. This allows us to select the spt3g dependent tests with `-m 'spt3g'`. We use this when running the tests on GitHub Actions.
- Updating the workflows to do the following:
  1. Install ocs
  2. Run non-spt3g, non-integration tests (this allows the CI to fail out early here if the faster unit tests don't pass)
  3. Build docker images
  4. Run spt3g dependent tests within the container
  5. Report coverage
  6. Publish images (if on develop/master branches)

Some other changes are included here that I noticed while testing. Those are:
- Adding Process running tracking to `_stop_process` tasks that were missing it. This results in returning True if a stop request was made while the Process was running, and False if a stop request was made when the Process wasn't running. This has been done elsewhere, but was missing from some of these early Agents.
- Removing a test build of the documentation from the GH Actions workflows. Like on socs, I've enabled PR builds on readthedocs, so that'll get tested in a more accurate environment.
- Updated the workflows to run on `ubuntu-latest`. I noticed the official release workflow was already on this, so I tested the workflows still on `ubuntu-18.04` on this. They worked, and I see no reason to stay on the older platform. This'll be `ubuntu-20.04` (until it isn't.)

## Motivation and Context

Just expanding our test coverage, working towards https://github.com/simonsobs/ocs/issues/28. This also provides examples for unit testing Agents.

## How Has This Been Tested?

```
$ python3 -m pytest --cov --cov-report=html -m 'not integtest'
==================================== test session starts =====================================
platform linux -- Python 3.9.7, pytest-6.2.5, py-1.10.0, pluggy-0.13.1
rootdir: /home/koopman/git/ocs/tests, configfile: pytest.ini
plugins: dependency-0.5.1, twisted-1.13.4, env-0.6.2, docker-compose-3.2.0, cov-2.10.1
collected 107 items / 5 deselected / 102 selected                                            

test_aggregator.py ..................                                                  [ 17%]
test_crossbar_integration.py .                                                         [ 18%]
test_influxdb_publisher.py ...                                                         [ 21%]
test_matched_client.py .                                                               [ 22%]
test_ocs_agent.py ............................                                         [ 50%]
test_ocs_feed.py ................                                                      [ 65%]
test_pacemaker.py ...                                                                  [ 68%]
test_rename.py ....                                                                    [ 72%]
test_site_config.py .....                                                              [ 77%]
agents/test_aggregator_agent.py .....                                                  [ 82%]
agents/test_fakedata.py ......                                                         [ 88%]
agents/test_host_master.py .                                                           [ 89%]
agents/test_influxdb_publisher_agent.py .....                                          [ 94%]
agents/test_ocs_plugin_standard.py .                                                   [ 95%]
agents/test_registry_agent.py .....                                                    [100%]

----------- coverage: platform linux, python 3.9.7-final-0 -----------
Coverage HTML written to dir htmlcov

======================= 102 passed, 5 deselected, 1 warning in 12.76s ========================
2021-10-14T20:56:04-0400 Main loop terminated.

```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

**Review by mhasself — APPROVED** [review]
Couple of comments inline.

Don't hold up merging on this, but my impressions from skimming the new agent tests are:
- I don't see many "assertRaises" (or whatever the fixture equivalent is) type tests -- it is important to fail loudly when the inputs or events are not valid.
- The motivations for the tests are not clear ... what important behavior of the agent is being tested?  This isn't always clear from the name of the function.

File: agents/aggregator/aggregator_agent.py
Comment (BrianJKoopman) [reply agents/aggregator/aggregator_agent.py:133]: Yeah, I was interested to hear thoughts on this. I agree it's a strange parameter. The reason I went this route is because the unit tests don't actually run the Agent, they run this method directly, and are caught in this loop without the ability to call stop.

**Comment by BrianJKoopman** [comment]
Thanks for taking a look, Matthew. Just to recap some points from our discussion of this on the telecon.

> My impressions from skimming the new agent tests are:
> 
>   * I don't see many "assertRaises" (or whatever the fixture equivalent is) type tests -- it is important to fail loudly when the inputs or events are not valid.

There aren't a lot of places that Agents raise exceptions. Generally we've been trying to catch exceptions that might come up, log them, and either move on or exit the Agent where necessary.

That isn't to say there are no cases where something will cause an Agent to raise an exception. I don't mean to state that these tests are exhaustive of all scenarios, but at least form a basis for adding tests to when such a scenario arises in the future.

>   * The motivations for the tests are not clear ... what important behavior of the agent is being tested?  This isn't always clear from the name of the function.

These tests are running the Tasks and Processes. It's a bit weird, because all a Task or Process ever directly returns is `True` or `False`. So when developing the tests I just tried to run each Task/Process with a test named something like `test_agent_name_task/process_name` and check that it completes successfully, i.e. `res[0] == True`. If that didn't execute all the code within a Task or Process I created a new test that covered whatever branch in the code was missed initially. 

Perhaps these could be better organized in the case that a given Task/Process requires multiple tests. Another thing that's occurring to me while writing this response is testing the contents of `session.data` for Tasks/Processes that use it.

As for test clarity -- I didn't write docstrings for a lot of these after looking around at how other projects tend to write unit tests (generally kept short and readable). These tests are all fairly short and I think somewhat readable. And while the integration testing in #231 does cover a lot of what's already covered here (by running Tasks/Processes), there is a lot more overhead there. This unit tests accomplish mostly the same thing (running the Task/Process) without needing to standup a crossbar server, copy of the Agent, etc, and thus run very quickly. My preference would be to write unit tests to cover as much as possible for an Agent, and cover remaining spots with integration testing. There will be some overlap here, but I think that's fine.

So, there are a few things I'll do before merging this:

- [ ] Group tests for a given Task/Process that requires multiple test into classes
- [ ] Test session.data contents where it's used
- [ ] Use the session information to determine if stop can be called properly, rather than some global like `self.aggregate`

**Comment by BrianJKoopman** [comment]
Thanks again for the feedback, all comments now addressed. I've rebased on the latest `develop` and will merge once checks are complete.

**Pull request merged**
